### PR TITLE
Debugger MemoryWidget: More display types, use combo box for options

### PR DESCRIPTION
**Pull request by TryTwo**
Added Int8/16/32 and double..

PR_display

- Do we want signed displays?
- 1.234e+09 syntax looks like it could be better, but what would I replace it with? Remove the +?
- Int8 looks terrible, but will look fine if we add a dual view  - two different column types with 4 bytes per row
- I left display type as a group, as it will likely get more items soon.
- Do we need to rename U8/16/32?
- Thoughts on sidebar rearrangement?

Most importantly, what else do I need to add? Do we need more customization in this PR like bytes per row?  A dual-view would require a lot of rearranging in the Update function to pass two view types through it. I think I would spin off column value updates and breakpoint tagging updates to two different functions, but will take suggestions on a better method to do it.

### CODE REVIEW

**Comment by TryTwo** [comment]
Added row length combo for 4 bytes, 8 bytes, or 16 bytes. Helps a lot with integer displays.

**Comment by AdmiralCurtiss** [comment]
The Search buttons should probably be closer to the top, probably right below the Set Value button, since they work of the same input field. Also get rid of the box they're in.

The display type might also look better with just a label next to it instead of an entire box for a single button.

**Comment by AdmiralCurtiss** [comment]
Rename U8/16/32 to Hex, probably. Might as well have Signed N and Unsigned N as separate options too.

Integers should be right-aligned within the column.

The Int32 view needs to be wide enough to fit the full integer, *please* don't use exponential for that, it's completely useless.

**Comment by TryTwo** [comment]
PR_display2

**Comment by AdmiralCurtiss** [comment]
...you should rebase this on master before you introduce even more conflicts.

**Comment by TryTwo** [comment]
Oh I didn't know it'd pick up rebase changes in Compare. That's a bit annoying.

**Comment by TryTwo** [comment]
I don't know how far this PR should go. I can break things up into two PRs, which is why I'm making multiple commits. Should I add this to this PR?
PR_display3

**Comment by AdmiralCurtiss** [comment]
Less changes per PR is better as long as they make sense individually -- easier to review and all. So yes, I'd split that up.

It's okay to have multiple PRs that depend on eachother, too, so you don't have to deal with merge conflicts. That is, one PR that has commit A only, and then another PR that has commits A and B with a note that the first PR should be merged before this one.

**Comment by TryTwo** [comment]
Ahh ok, that makes sense.

**Comment by TryTwo** [comment]
Made clicking preserve byte offset (like 80000001), made row breakpoint conform to row settings.

Breakpointing odd bytes (0x1) has weird results, I'm not sure if I need to enforce some sort of even byte check. Also, a 4-byte row breakpoint on a 0x2 offset is weird.

**Comment by TryTwo** [comment]
@AdmiralCurtiss  I need your thoughts on how the rows should be aligned if you input 80000001 or 80000002.  Keep in mind integer values change if the alignment isn't a multiple of the underlying byte length.  The easy thing to do is forcibly round the row addresses to 0 at all times and just highlight the target address so you know where to look, but I'm thinking you need more functionality than that.

4 byte and 8 byte row lengths would of course if you rows aligned like that: Like 80000000, 80000004, 80000008 for byte.

**Comment by sepalani** [comment]
> @AdmiralCurtiss I need your thoughts on how the rows should be aligned if you input 80000001 or 80000002. Keep in mind integer values change if the alignment isn't a multiple of the underlying byte length. The easy thing to do is forcibly round the row addresses to 0 at all times and just highlight the target address so you know where to look, but I'm thinking you need more functionality than that.
> 
> 4 byte and 8 byte row lengths would of course if you rows aligned like that: Like 80000000, 80000004, 80000008 for byte.

What about adding a setting for that and let the user set the alignment such as: _(feel free to come with better names/wording)_
 - auto: alignment based on type and the number of bytes per row
 - none: will go at the address+offset without taking the alignment into account
 - fixed: based on the number of bytes per row (the behaviour of hex editor)
 - custom: based on the user input

I'm fine with moving the memory dump buttons down below as long as they are in a dedicated group box like the other options.

File: Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp [hunk 1]
```diff
@@ -63,28 +63,28 @@ void MemoryViewWidget::UpdateFont()
   Update();
 }
 
-static int GetColumnCount(MemoryViewWidget::Type type)
+static int GetByteLength(MemoryViewWidget::Type type)
```
Comment (sepalani) [inline Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:66]: A `byte` or rather an [octet](https://en.wikipedia.org/wiki/Byte) is always 8 bits so "byte length" doesn't really make sense in this context. A better wording would be type length or type size, I guess.

Comment (TryTwo) [reply Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:66]: ByteCount? I think it needs the word byte in there to make the value returned obvious, as bits are referenced often..

**Comment by TryTwo** [comment]
> What about adding a setting for that and let the user set the alignment such as: _(feel free to come with better names/wording)_

That might work. Would it be better to put that in the table's context menu?

File: Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp (hunk 1 above)
Comment (sepalani) [reply Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:66]: Often in C/C++ you'll find `size`/`length`/`sizeof` referring to the size in bytes. Wording is hard and I don't I have a better name on top of my head right now...

**Comment by AdmiralCurtiss** [comment]
I think by default just aligning starting at the current address is fine, if you jump to ...02 then you'll just get 02, 06, 0a, 0e when 4-byte aligned. If you want to be fancy you can make options, but I dunno if it's really all that necessary -- you can just go back 2 bytes after all.

**Comment by TryTwo** [comment]
Alright, how is that for options?

Too many things were being called address / addr, so I renamed some things, and then I renamed the associated things as well.

File: Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp (hunk 1 above)
Comment (TryTwo) [reply Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:79]: Oops, I forgot to do this, I'll include it in the next update.

**Comment by AdmiralCurtiss** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp [hunk 2]
```diff
@@ -191,18 +251,45 @@ void MemoryViewWidget::Update()
...
+    case Type::Signed16:
+      update_values([&accessors](u32 address) {
+        return QString::number(Common::BitCast<signed short>(accessors->ReadU16(address)));
+      });
+      break;
+    case Type::Signed32:
+      update_values([&accessors](u32 address) {
+        return QString::number(Common::BitCast<signed int>(accessors->ReadU32(address)));
+      });
+      break;
```
Comment (AdmiralCurtiss) [inline Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:292]: You can and should probably use `s8`/`s16`/`s32` here for the casts.

File: Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp [hunk 3]
```diff
@@ -35,6 +36,19 @@ MemoryViewWidget::MemoryViewWidget(QWidget* parent) : QTableWidget(parent)
 
   setContextMenuPolicy(Qt::CustomContextMenu);
 
+  // Context menu actions.
+  m_align_fixed = new QAction(tr("Fixed"));  //, this, [this] { SetAddress(m_address); });
+  m_align_auto = new QAction(tr("Auto"));    //, this, [this] { SetAddress(m_address); });
+  m_align_none = new QAction(tr("None"));    //, this, [this] { SetAddress(m_address); });
```
Comment (AdmiralCurtiss) [inline Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:42]: What's up with this commented out code?

Comment (TryTwo) [reply Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:42]: Ah sorry I left a bit of trash

File: Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp (hunk 2 above)
Comment (TryTwo) [reply Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:292]: Oh I didn't know you could use that. Thanks

**Comment by TryTwo** [comment]
The only thing left out was a custom alignment, as I didn't want to make an input box for it. If needed, I could do a "lock alignment" option, and then the address box would be used to set the alignment, then after it is locked, it wouldn't change even if the address does.

**Comment by TryTwo** [comment]
Should be pretty much done with the PR, unless @[sepalani](https://github.com/sepalani) wants more for alignment, or if anyone has other thoughts.

For the UI buttons, search has a label that pops up giving you feedback, so I left the box alone. Then since everything else has a box, dump got one for consistency.

**Comment by Dentomologist** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/DolphinQt/Debugger/MemoryWidget.cpp [hunk 4]
```diff
@@ -179,22 +186,32 @@ void MemoryWidget::CreateWidgets()
...
+  displaytype_group->setLayout(displaytype_layout);
+
+  m_display_combo = new QComboBox;
+  m_display_combo->setMaxVisibleItems(20);
+  m_display_combo->addItem(tr("Hex 8"), int(MemoryViewWidget::Type::Hex8));
+  m_display_combo->addItem(tr("Hex 16"), int(MemoryViewWidget::Type::Hex16));
+  m_display_combo->addItem(tr("Hex 32"), int(MemoryViewWidget::Type::Hex32));
+  m_display_combo->addItem(tr("Unsigned 8"), int(MemoryViewWidget::Type::Unsigned8));
+  m_display_combo->addItem(tr("Unsigned 16"), int(MemoryViewWidget::Type::Unsigned16));
+  m_display_combo->addItem(tr("unsigned 32"), int(MemoryViewWidget::Type::Unsigned32));
```
Comment (AdmiralCurtiss) [inline Source/Core/DolphinQt/Debugger/MemoryWidget.cpp:200]: This string is randomly lowercase.

File: Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp [hunk 5]
```diff
@@ -62,28 +76,76 @@ void MemoryViewWidget::UpdateFont()
...
   clearSelection();
 
-  setColumnCount(2 + GetColumnCount(m_type));
+  // Optional Alignments, always a power of 2.
+  u32 address = m_address;
+
+  if (m_align_fixed->isChecked())
+    address = address & -m_bytes_per_row;
+  else if (m_align_auto->isChecked())
+    address = address & -GetTypeSize(m_type);
```
Comment (AdmiralCurtiss) [inline Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:144]: I see what you're doing here, but some compilers are very likely to warn about taking a negative of an unsigned here. You probably want to write this as `address & ~(m_bytes_per_row - 1)` instead. Same for the GetTypeSize() one.

File: Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp [hunk 6]
```diff
@@ -102,83 +164,81 @@ void MemoryViewWidget::Update()
...
-    if (m_address_space == AddressSpace::Type::Effective)
-    {
-      auto* description_item = new QTableWidgetItem(
-          QString::fromStdString(PowerPC::debug_interface.GetDescription(addr)));
-
-      description_item->setForeground(Qt::blue);
-      description_item->setFlags(Qt::ItemIsEnabled);
-
-      setItem(i, columnCount() - 1, description_item);
-    }
```
Comment (AdmiralCurtiss) [inline Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:146]: Why did you remove this?

File: Source/Core/DolphinQt/Debugger/MemoryWidget.cpp [hunk 7]
```diff
@@ -179,22 +186,32 @@ void MemoryWidget::CreateWidgets()
...
-  datatype_layout->addWidget(m_type_u32);
-  datatype_layout->addWidget(m_type_ascii);
-  datatype_layout->addWidget(m_type_float);
-  datatype_layout->setSpacing(1);
+  auto* displaytype_group = new QGroupBox(tr("Display Type"));
+  auto* displaytype_layout = new QVBoxLayout;
+  displaytype_group->setLayout(displaytype_layout);
+
+  m_display_combo = new QComboBox;
+  m_display_combo->setMaxVisibleItems(20);
```
Comment (AdmiralCurtiss) [inline Source/Core/DolphinQt/Debugger/MemoryWidget.cpp:194]: Is this necessary?

**Review by AdmiralCurtiss — COMMENTED** [review]
I feel like the alignment option should be another combobox or something by the Display Type options, rather than hidden inside the right-click menu. The names also aren't very self-explanatory, but to be fair I don't have a good suggestion either and it's easy enough to figure out with some experimentation. Otherwise this seems mostly fine.

File: Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp (hunk 5 above)
Comment (AdmiralCurtiss) [reply Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:144]: Actually, I just remembered we have a `Common::AlignDown()` function, maybe use that.

Comment (TryTwo) [reply Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:144]: Thanks, I was trying to find a support function

File: Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp (hunk 6 above)
Comment (TryTwo) [reply Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:146]: It  places symbol data in a column that is overwritten by normal value data.  Am I missing some case where it actually appears?

File: Source/Core/DolphinQt/Debugger/MemoryWidget.cpp (hunk 7 above)
Comment (TryTwo) [reply Source/Core/DolphinQt/Debugger/MemoryWidget.cpp:194]: The combobox will have a scroll bar if we don't increase the size. I think it's better to just see all the options.

**Comment by TryTwo** [comment]
Made the alignment combobox.  Also squashed the two commits to one, it was too hard to keep them split. Not sure if "No Alignment" should have a warning or be disabled for floats, doubles, and maybe ints, as the value changes.

**Comment by dreamsyntax** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp (hunk 6 above)
Comment (AdmiralCurtiss) [reply Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:146]: Mmm you're right but this is clearly trying to do something useful. Whatever, we can fix it later.

**Comment by AdmiralCurtiss** [comment]
Alright, lemme test this again but this seems fine to me now.

**Comment by AdmiralCurtiss** [comment]
You didn't fix the random lowercase `unsigned`! But yeah seems good otherwise.

File: Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp (hunk 6 above)
Comment (TryTwo) [reply Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:146]: Maybe two column mode can have address + symbol.  The symbols are what you get from codewidget.  They could also be used as bookmarks if there were additional UI stuff to make and edit them in memorywidget.

**Comment by TryTwo** [comment]
I got it this time.

**Comment by sepalani** [comment]
@dolphin-emu-bot rebuild
